namespace/bcl.adoc

```diff
@@ -0,0 +1,46 @@
+= Binary Compatibility Levels
+
+This is a classification of the levels of binary compatibility support for Java EE 8 classic applications in terms of the level of difficulty/complexity of the associated implementation and tools.
+
+== Level 0 - Jakarta EE 8 Only
```

Comment:
I don't understand this level.  Is this a product that supports only Jakarta EE 8?
Obviously the binary compatibility with Jakarta EE 8 for such a product would
be "complete".  There would be no issue of backwards compatibility.  There
would be no choice for implementers to decide.

Obviously you must mean something else by this level but I don't understand
what it is.

Reply:
Bill, both Scott and I identified a mode 0 as just running an EE8 container/server as a way to support EE8 applications.   Yes it is kind of obvious to us, but I think it is important for us to say as much to the community - ie that we will continue supporting our EE8 releases for sometime and that there is no immediate prerogative to upgrade to an EE9 capable container.

Reply:
I think it just confuses the discussion of whether or how to provide compatibility
with Jakarta EE 8 in a Jakarta EE 9 or later product.

Reply:
I think it precisely defines how many vendors are going to provide compatibility for Jakarta EE8, once EE9 is released.  We are going to maintain our pure EE8 platforms for a lot longer than a normal version change because there will be some applications that cannot or willnot change (or be changeable) to the new APIs.   With APIs like servlet-api, many version increments have been very minor and thus there has been very little impediment to the end-of-life of previous versions as binary compatibility was trivial.  This will not be the case and vendors are going to have to decide which of the other modes - if any - they will implement.   Mode 0, will be the primary mode for many, perhaps the only mode.

Reply:
That's not providing compatibility.  That's like saying Java is compatible with C# because I can tell you to go use that C# product to run your C# application.

Reply:
To run with your analogy, mode zero is saying to C# vendors that they may well be moving to become Go runtime vendors, but they have to keep maintaining their C# runtimes until such time is that fancy new Go runtime can also seamlessly run C# applications.

Reply:
We're not telling vendors what they *have* to do.  At this point, we're *not* expecting
that compatibility is *required*.

So yes, the C# runtime vendor can abandon all their C# customers if they think that's
a better business model, or they can tell them that there's no path for them to take
advantage of new features or performance in the vendor's new Go runtime product so
they should keep using the C# product, or they can build a Go runtime product that provides
some compatibility for their C# applications and allows the customers to benefit from the
new features and performance in the vendor's Go runtime product.  It's the vendor's choice.

Reply:
Exactly, there are lots of options and this document is about enumerating them.    So perhaps there is also a mode -1,  so using my documents numbering:
 + mode -1 - no EE8 support, EE9 only
 + mode 0 - separate EE8 and EE9 servers
 + mode 1 - isolated EE8 and EE9 containers in the same server
 + mode 2 - EE8 and EE9 containers in same server with shared JNDI space
 + mode 3 - EE8 wrapped as EE9 in a EE9 only server
 + mode 4 - EE8 translated to EE9 in EE9 only server
 + mode 5 - magical mixed EE8 and EE9 container.

I see no harm in enumerating all the possibilities, and I see benefit in being clear in calling out that mode 0 is a likely outcome/step for many vendors.  We are not telling vendors which modes they must implement, but in discussing how our EE ecosystem moves forward it is important to have a complete picture.

Reply:
Maybe this is a matter of perspective...

For me, the important issue is what binary compatibility does a Jakarta EE 9 **product**
support.  You seem to be answering the question of what binary compatibility does a
Jakarta EE 9 **vendor** support.  The former is a technical problem.  The latter can be
dealt with as just a business issue.

From a vendor perspective, the choices then become...

- mode -1 - vendor provides only an EE 9 product
- mode 0 - vendor provides separate EE 8 and EE 9 products
- mode 1 - vendor provides a single product that includes isolated EE 8 and EE 9 containers

and so on.  In my mind, mode -1 and mode 0 are "no binary compatibility".
I'm much more interested in the technical choices for how **to** provide binary compatibility.

Reply:
Bill, I totally agree that modes -1 and 0 are non technical modes that are "no binary compatibility".    Also agree that it is more important to work out technical details of modes >0.
I just think that from a project point of view it is important to enumerate all the possibilities and "no binary compatibility" is a valid possibility.     Anyway... I think we've established we have different points of view on this one... let others adjudicate and let's focus on the technical details of modes >0